pyphare/pyphare_tests/test_pharesee/test_geometry.py

```diff
@@ -0,0 +1,528 @@
+                        {"box": Box(64, 65),"offset":(65 ,0)}
+                    ],
+
+            1:  # level 1
+                [
+                    {"box": Box(59, 65),'offset': (0, 0)},
+                    {"box": Box(59, 64),"offset": (0, 0)},
+                    {"box": Box(59, 64),"offset": (0, 0)},
+                    {"box": Box(59, 64),'offset': (0, 0)},
+                    {"box": Box(59, 65),"offset": (0, 0)},
+                    {"box": Box(59, 65),"offset": (0, 0)}
+                ]
+
+
+        }
+
+
+        overlaps = hierarchy_overlaps(self.hierarchy)
+
+        for ilvl, lvl in enumerate(self.hierarchy.patch_levels):
+            self.assertEqual(len(expected[ilvl]), len(overlaps[ilvl]))
+
+            for exp, actual in zip(expected[ilvl], overlaps[ilvl]):
+
+                act_box    = actual["box"]
+                act_offset = actual["offset"]
+                exp_box    = exp["box"]
+                exp_offset = exp["offset"]
+
+                self.assertEqual(act_box, exp_box)
+                self.assertEqual(act_offset, exp_offset)
+
+
+
+
+    def test_touch_border(self):
+
+        self.assertFalse(touch_domain_border(Box(10,20), self.hierarchy.domain_box, "upper"))
+        self.assertFalse(touch_domain_border(Box(10, 20), self.hierarchy.domain_box, "lower"))
+        self.assertTrue(touch_domain_border(Box(0, 20), self.hierarchy.domain_box, "lower"))
+        self.assertTrue(touch_domain_border(Box(-5, 20), self.hierarchy.domain_box, "lower"))
+        self.assertTrue(touch_domain_border(Box(-5, 70), self.hierarchy.domain_box, "lower"))
+        self.assertTrue(touch_domain_border(Box(-5, 70), self.hierarchy.domain_box, "upper"))
+        self.assertTrue(touch_domain_border(Box(40, 70), self.hierarchy.domain_box, "upper"))
+        self.assertTrue(touch_domain_border(Box(40, 64), self.hierarchy.domain_box, "upper"))
+
+
+
+
+    def test_particle_ghost_area_boxes(self):
+
+        expected = {
+
+            0: [ {"boxes":[Box(33, 33), Box(-1,-1)]},
+                 {"boxes": [Box(32, 32), Box(65, 65)]} ],
+
+
+            1: [ {"boxes": [Box(9, 9), Box(60, 60)]},
+                 {"boxes": [Box(63, 63), Box(112, 112)]} ]
+        }
+
+        gaboxes = particle_ghost_area_boxes(self.hierarchy)
+
+        # same number of levels
+        self.assertEqual(len(expected), len(gaboxes))
+
+        for ilvl, lvl in enumerate(self.hierarchy.patch_levels):
+
+            # same number of PatchDatas
+            self.assertEqual(len(gaboxes[ilvl]), len(expected[ilvl]))
+
+            for act_pdata, exp_pdata in zip(gaboxes[ilvl], expected[ilvl]):
+
+                    self.assertEqual(len(exp_pdata["boxes"]), len(act_pdata["boxes"]))
+
+                    for exp_box in exp_pdata["boxes"]:
+                        self.assertTrue(exp_box in act_pdata["boxes"])
+
+
+
+
+
+    def test_level_ghost_boxes(self):
+
+        expected = {
+
+            1: [
+                {"boxes": [Box(9, 9),]},
```

Comment:
I still do not understand these boxes : I thought their size was lower, upper, so how can a box of width != 0 with lower=upper ?

Reply:
lower==upper means the box is 1 cell wide